### PR TITLE
magicsock: some tests

### PR DESCRIPTION
**Pull request by crawshaw**
Hopefully more coming later today. Current TODO list:

- [x] add DERP
  - [x] test several hundred packets make it through
  - [ ] test spray works as intended
- [x] add DERP and make one normal ep invalid
- [x] add DERP and make both normal eps invalid
- [ ] try and verify re-STUN
- [ ] introduce latency so that DERP wins 50% of the time.

### CODE REVIEW

File: wgengine/magicsock/magicsock_test.go [hunk 1]
```diff
@@ -77,3 +78,58 @@ func TestDerpIPConstant(t *testing.T) {
...
+	t.Helper()
+
+	// TODO(crawshaw): use stats to test re-STUN logic
+	var stats stunStats
+
+	pc, err := net.ListenPacket("udp4", ":3478")
+	if err != nil {
+		t.Fatalf("failed to open STUN listener: %v", err)
+	}
+	t.Cleanup(func() { pc.Close() })
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock_test.go]: We officially switching to requiring Go 1.14 now? (Fine with me, but we need to update the CI first.)

File: wgengine/magicsock/magicsock_test.go [hunk 2]
```diff
@@ -77,3 +78,58 @@ func TestDerpIPConstant(t *testing.T) {
...
+	go runSTUN(pc, &stats)
+	return pc.LocalAddr()
+}
+
+func runSTUN(pc net.PacketConn, stats *stunStats) {
+	var buf [64 << 10]byte
+	for {
+		n, addr, err := pc.ReadFrom(buf[:])
+		if err != nil {
+			continue
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock_test.go]: do you mean return? When does this ever exit?

File: derp/derphttp/derphttp_client.go [hunk 3]
```diff
@@ -30,6 +30,9 @@ import (
 	"tailscale.com/types/logger"
 )
 
+// DefaultTLSConfig is the base *tls.Config used by a DERP Client.
+var DefaultTLSConfig = &tls.Config{}
```
Comment (bradfitz) [inline derp/derphttp/derphttp_client.go]: I'd prefer if this were moved to an optional exported field on Client, just named TLSConfig. If nil, automatic.

**Comment by crawshaw** [comment]
Sounds good. I'm just pushing to checkpoint in case my laptop gets soaked, will address your comments tonight. (Also this test found another data race.)

File: wgengine/magicsock/magicsock_test.go (hunk 1 above)
Comment (crawshaw) [reply wgengine/magicsock/magicsock_test.go]: Will do later (too much going on for me to switch it now).

File: derp/derphttp/derphttp_client.go [hunk 4]
```diff
@@ -37,6 +37,8 @@ import (
 // Send/Recv will completely re-establish the connection (unless Close
 // has been called).
 type Client struct {
+	TLSConfig *tls.Config
```
Comment (bradfitz) [inline derp/derphttp/derphttp_client.go]: document this. I assume this is optional and nil means automatic?

File: wgengine/magicsock/magicsock.go [hunk 5]
```diff
@@ -78,12 +79,13 @@ type Conn struct {
...
-	derpConn    map[int]*derphttp.Client   // magic derp port (see derpmap.go) to its client
-	derpCancel  map[int]context.CancelFunc // to close derp goroutines
-	derpWriteCh map[int]chan<- derpWriteRequest
+	derpMu        sync.Mutex
+	privateKey    key.Private
+	myDerp        int                        // nearest DERP server; 0 means none/unknown
+	derpConn      map[int]*derphttp.Client   // magic derp port (see derpmap.go) to its client
+	derpCancel    map[int]context.CancelFunc // to close derp goroutines
+	derpWriteCh   map[int]chan<- derpWriteRequest
+	derpTLSConfig *tls.Config
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock.go]: Add same-line comment `// normally nil; used by tests`

**Pull request merged**